fiftyone/core/collections.py

```diff
@@ -130,6 +161,20 @@ def __init__(
         self._encoding_ratio = 1.0
         self._last_time = None
 
+        self.samples_lock = threading.Lock()
+        self.frames_lock = threading.Lock()
+        self.batch_ids_lock = threading.Lock()
+        self.reloading_lock = threading.Lock()
+
+        self.executor = (
+            # Using more than one worker will introduce race conditions in the state preserved between DB writes, but
+            # since prediction is the bottleneck, not writing to the DB, only one worker is necessary.
```

Comment:
I think save context is used in more than just prediction-related tasks. Iirc, it's in iter_samples as well